tests/test_information_from_events.py

```diff
@@ -0,0 +1,464 @@
+            (event["blockNumber"], event["logIndex"], event["transactionHash"]),
+            event["args"]["_value"],
+        )
+        for event in balance_update_events
+    ]
+    balance_updates.extend(
+        (
+            (event["blockNumber"], event["logIndex"], event["transactionHash"]),
+            -event["args"]["_value"],
+        )
+        for event in reverse_balance_update_events
+    )
+    balance_updates.sort()
+    return balance_updates
+
+
+def get_transfer_balance_update_events(currency_network_contract, transfer_event):
+    """Returns all balance update events in the correct order that belongs to the transfer event"""
+    log_index = transfer_event["logIndex"]
+    tx_hash = transfer_event["transactionHash"]
+
+    receipt = currency_network_contract.web3.eth.getTransactionReceipt(tx_hash)
+    balance_update_events = currency_network_contract.events.BalanceUpdate().processReceipt(
+        receipt
+    )
+
+    sender = transfer_event["args"]["_from"]
+    receiver = transfer_event["args"]["_to"]
+
+    balance_events = []
+    saw_sender_event = False
+    saw_receiver_event = False
+
+    # Search backwards for releated BalanceUpdate events
+    for i in range(log_index - 1, -1, -1):
+        for event in balance_update_events:
+            if event["logIndex"] == i:
+                balance_events.append(event)
+                if event["args"]["_to"] == receiver:
+                    saw_receiver_event = True
+                if event["args"]["_from"] == sender:
+                    saw_sender_event = True
+                break
+        if saw_sender_event and saw_receiver_event:
+            break
+    else:
+        assert False, "Could not find all BalanceUpdate events"
+
+    if balance_events[0]["args"]["_from"] != sender:
+        # For the sender pays case, they are reverse
+        balance_events.reverse()
+
+    assert balance_events[0]["args"]["_from"] == sender
+    assert balance_events[-1]["args"]["_to"] == receiver
+    return balance_events
+
+
+def get_transfer_path(currency_network_contract, transfer_event):
+    """Returns the transfer path of the given transfer without the sender"""
+    path_from_events = []
+    for event in get_transfer_balance_update_events(
+        currency_network_contract, transfer_event
+    ):
+        path_from_events.append(event["args"]["_to"])
+
+    return path_from_events
+
+
+def get_previous_balance(balance_updates, balance_update_event):
+    """Returns the balance before a given balance update event"""
+    index = 0
+    # find the corresponding event
+    for i, (key, value) in enumerate(balance_updates):
+        if key == (
+            balance_update_event["blockNumber"],
+            balance_update_event["logIndex"],
+            balance_update_event["transactionHash"],
+        ):
+            index = i
+            break
+    else:
+        raise RuntimeError("Could not find balance update")
+    index -= 1
+
+    if index < 0:
+        return 0
```

Comment:
I guess you assume that if the index of `balance_update_event` within `balance_updates` is 0 then it is the first ever `BalanceUpdate` event and the previous balance was 0. But it could also be that the caller gave a truncated list of `balance_updates` where the sought event has index 0 and the previous one is not in the list (even though it exists).

I would rather we get the `balance_updates` from within the function to enforce getting all of them, or at least document that all events from block 0 should be given (by function documentation or argument name).

Reply:
Get it now within the function